File: WdtcCore/src/main/java/org/wdt/wdtc/launch/AbstractGameCommand.java
package org.wdt.wdtc.launch;

import org.wdt.wdtc.utils.StringUtils;

import java.io.File;
import java.io.IOException;

abstract class AbstractGameCommand {
    protected final StringBuilder Command;

    public AbstractGameCommand() {
        Command = new StringBuilder();
    }

    abstract StringBuilder getCommand() throws IOException;

    protected void NonBreakingSpace(Object o) {
        Command.append(o).append(StringUtils.STRING_SPACE);
    }

    protected void InsertclasspathSeparator(File file) {
        InsertclasspathSeparator(file.getAbsolutePath());
    }

    protected void InsertclasspathSeparator(String str) {
        Command.append(str).append(";");
    }

    protected void InsertSpace(String str) {
        Command.append(str).append(StringUtils.STRING_SPACE);
    }

    protected void NonBreakingSpace(String str, Object string) {
        NonBreakingSpace(str + string);
    }
}

File: core/core-impl/src/main/kotlin/org/wdt/wdtc/core/impl/launch/AbstractGameCommand.kt
package org.wdt.wdtc.core.impl.launch

import org.wdt.wdtc.core.openapi.utils.STRING_SPACE


abstract class AbstractGameCommand {
  protected val commandBuilder: StringBuilder = StringBuilder()
  private val regex = ".*\\\$\\{(.*)}".toRegex()

  abstract fun getCommand(): StringBuilder

  abstract val dataMap: Map<String, String>
	
	protected fun nonBreakingSpace(o: Any) {
    commandBuilder.append(o).append(STRING_SPACE)
  }

  protected fun nonBreakingSpace(str: String, o: Any) {
    commandBuilder.append(str).append(o).append(STRING_SPACE)
  }

  protected fun String.replaceData(dataMap: Map<String, String>): String {
    if (this.matches(regex)) {
      return regex.find(this)?.groups?.get(1)?.value.let { str ->
        dataMap[str].let {
          if (it == null) this
          else replace("\${$str}", it)
        }
      }
    }
    return this
  }
}
